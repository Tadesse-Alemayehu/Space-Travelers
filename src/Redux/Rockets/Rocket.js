const STORE_ROCKETS = 'src/pages/rockets/ADD_ROCKETS';
const TOGGLE_RESERVATION = 'src/pages/rockets/TOGGLE_RESERVATION';
export const toggleReservedAction = (id, status) => ({ type: STORE_ROCKETS, payload: { status, id } });
export const addRockets = () => (dispatch) => {
  fetch('https://api.spacexdata.com/v3/rockets')
    .then(data => data.json())
    .then(data => {
      console.log(data);
      dispatch({ type: STORE_ROCKETS, payload: data });
    })
    .catch(error => console.log(error));
};

const reduceBook = (state = [], action) => {
  switch (action.type) {
    case STORE_ROCKETS:
      return action.payload;
    case TOGGLE_RESERVATION:
      return [...state.map(rocket => {
        if (rocket.id === action.payload.id)
          return { ...rocket, reserved: action.payload.status }
        return rocket;
      }
      )];
    default:
      return state;
  }
}


export default reduceBook;

