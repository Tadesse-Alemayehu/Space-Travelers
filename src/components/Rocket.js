import React from 'react';
import { Row, Col, Button } from 'react-bootstrap'
import 'bootstrap/dist/css/bootstrap.min.css';
export default function Rocket({ id, name, description, image, reserved }) {
  return (
    <Row className="rocket-container align-items-center mt-md-3">
      <Col sm={5} md={4} lg={3}>
        <img src={image} alt={name} width="100%" />
      </Col>
      <Col className="rocket-description">
        <h3 className="text-capitalize">{name}</h3>
        <p>{description}</p>
        {reserved ?
          (<Button variant="outline-dark " className="reserved text-capitalize" id={id} data-reserved={false} >cancel reservation</Button>) :
          (<Button className="not-reserved text-capitalize" id={id} data-reserved={true} >reserve rocket</Button>)}
      </Col>
    </Row>
  )
}