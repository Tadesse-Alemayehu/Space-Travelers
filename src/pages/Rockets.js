import React from 'react';
import { useSelector } from 'react-redux';
import Container from 'react-bootstrap/container';
import Rocket from '../components/Rocket';

export default function Rockets() {
  const rockets = useSelector((state) => state.Rockets);
  return (
    <div style={{ marginBottom: '5vw' }}>
      <Container data-testid="list-of-rockets">
        {rockets.map((rocket) => (
          <Rocket
            key={rocket.id}
            id={rocket.id}
            name={rocket.rocket_name}
            description={rocket.description}
            image={rocket.flickr_images[0]}
            reserved={rocket.reserved}
          />
        ))}
      </Container>
    </div>
  );
}
