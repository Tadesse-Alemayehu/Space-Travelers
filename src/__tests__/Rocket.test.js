import { render, screen } from '@testing-library/react';
import Rocket from '../components/Rocket';
import { rocketData, STORE_ROCKETS } from '../__mocks__/MockRocketData.js';
// import { toggleReservedAction } from '../Redux/Rockets/Rocket'
import { Provider } from 'react-redux';
import Store from '../Redux/configureStore';
import * as RocketActions from '../Redux/Rockets/Rocket'

const mockRocketAPICall = jest.spyOn(RocketActions, 'addRockets')
  .mockImplementation(() => {
    console.log("diapatching action");
    return { type: STORE_ROCKETS, payload: rocketData };
  });
describe('Render the Rocket component and check for Proper Rocket component', () => {
  Store.dispatch(RocketActions.addRockets());
  render(
    <Provider store={Store}>
      {rocketData.map((rocket) =>
        <Rocket
          key={rocket.id}
          id={rocket.id}
          name={rocket.rocket_name}
          description={rocket.description}
          image={rocket.flickr_images[0]}
          reserved={rocket.reserved}
        />)}
    </Provider>
  );

  test('Assert store is updated with a mock data', () => {
    expect(Store.getState().Rockets.length).toBe(rocketData.length)
    // console.log(Store.getState().Rockets.length)
  })

});

afterEach(() => {
  jest.clearAllMocks();
});
