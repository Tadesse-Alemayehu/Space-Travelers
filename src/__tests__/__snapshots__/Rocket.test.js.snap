// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snap shot tests for Rocket Test snapshot match 1`] = `
Array [
  <div
    className="rocket-container align-items-center mt-md-3 row"
    data-testid="rocket"
  >
    <div
      className="col-lg-3 col-md-4 col-sm-5"
    >
      <img
        alt="Falcon 1"
        src="https://imgur.com/DaCfMsj.jpg"
        width="100%"
      />
    </div>
    <div
      className="rocket-description col"
    >
      <h3
        className="text-capitalize"
      >
        Falcon 1
      </h3>
      <p>
        The Falcon 1 was an expendable launch system privately developed and manufactured by SpaceX during 2006-2009. On 28 September 2008, Falcon 1 became the first privately-developed liquid-fuel launch vehicle to go into orbit around the Earth.
      </p>
      <button
        className="not-reserved text-capitalize btn btn-primary"
        data-id={1}
        data-reserved={1}
        disabled={false}
        id={1}
        onClick={[Function]}
        type="button"
      >
        reserve rocket
      </button>
    </div>
  </div>,
  <div
    className="rocket-container align-items-center mt-md-3 row"
    data-testid="rocket"
  >
    <div
      className="col-lg-3 col-md-4 col-sm-5"
    >
      <img
        alt="Falcon 9"
        src="https://farm1.staticflickr.com/929/28787338307_3453a11a77_b.jpg"
        width="100%"
      />
    </div>
    <div
      className="rocket-description col"
    >
      <h3
        className="text-capitalize"
      >
        Falcon 9
      </h3>
      <p>
        Falcon 9 is a two-stage rocket designed and manufactured by SpaceX for the reliable and safe transport of satellites and the Dragon spacecraft into orbit.
      </p>
      <button
        className="not-reserved text-capitalize btn btn-primary"
        data-id={2}
        data-reserved={1}
        disabled={false}
        id={2}
        onClick={[Function]}
        type="button"
      >
        reserve rocket
      </button>
    </div>
  </div>,
  <div
    className="rocket-container align-items-center mt-md-3 row"
    data-testid="rocket"
  >
    <div
      className="col-lg-3 col-md-4 col-sm-5"
    >
      <img
        alt="Falcon Heavy"
        src="https://farm5.staticflickr.com/4599/38583829295_581f34dd84_b.jpg"
        width="100%"
      />
    </div>
    <div
      className="rocket-description col"
    >
      <h3
        className="text-capitalize"
      >
        Falcon Heavy
      </h3>
      <p>
        With the ability to lift into orbit over 54 metric tons (119,000 lb)--a mass equivalent to a 737 jetliner loaded with passengers, crew, luggage and fuel--Falcon Heavy can lift more than twice the payload of the next closest operational vehicle, the Delta IV Heavy, at one-third the cost.
      </p>
      <button
        className="not-reserved text-capitalize btn btn-primary"
        data-id={3}
        data-reserved={1}
        disabled={false}
        id={3}
        onClick={[Function]}
        type="button"
      >
        reserve rocket
      </button>
    </div>
  </div>,
  <div
    className="rocket-container align-items-center mt-md-3 row"
    data-testid="rocket"
  >
    <div
      className="col-lg-3 col-md-4 col-sm-5"
    >
      <img
        alt="Starship"
        src="https://live.staticflickr.com/65535/48954138962_ee541e6755_b.jpg"
        width="100%"
      />
    </div>
    <div
      className="rocket-description col"
    >
      <h3
        className="text-capitalize"
      >
        Starship
      </h3>
      <p>
        Starship and Super Heavy Rocket represent a fully reusable transportation system designed to service all Earth orbit needs as well as the Moon and Mars. This two-stage vehicle — composed of the Super Heavy rocket (booster) and Starship (ship) — will eventually replace Falcon 9, Falcon Heavy and Dragon.
      </p>
      <button
        className="not-reserved text-capitalize btn btn-primary"
        data-id={4}
        data-reserved={1}
        disabled={false}
        id={4}
        onClick={[Function]}
        type="button"
      >
        reserve rocket
      </button>
    </div>
  </div>,
]
`;
